github/resource_github_repository_ruleset.go

```diff
@@ -0,0 +1,404 @@
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the repository to add the ruleset to.",
+			},
+			"owner": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "Owner of the repository. If not provided, the provider's default owner is used.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				Description: "The name of the Ruleset within the repository.",
+			},
+			"target": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				Description: "The target of the ruleset. Either branch or tag.",
+				ValidateFunc: validation.StringInSlice([]string{
+					"branch",
+					"tag",
+				}, false),
+			},
+			"enforcement": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				Description: "The enforcement level of the ruleset. One of active, disabled or evaluate. `evaluate` allows admins to test rules before enforcing them. Admins can view insights on the Rule Insights page (`evaluate` is only available with GitHub Enterprise).",
+				ValidateFunc: validation.StringInSlice([]string{
+					"active",
+					"disabled",
+					"evaluate",
+				}, false),
+			},
+			"conditions": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				MaxItems:    1,
+				Description: "Target branches/tags. Both an inclusion and exclusion list, supporting regexes as well as ALL branches/tags and the default branch",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						// TODO: Should the default branch + ALL branches/tags have it's own field?
+						"include": {
+							Type:        schema.TypeSet,
+							Optional:    true,
+							ForceNew:    true, // TODO: Remove this when updating is implemented
+							Description: "Array of ref names or patterns to include. One of these patterns must match for the condition to pass. Also accepts `~DEFAULT_BRANCH` to include the default branch or `~ALL` to include all branches.",
+							Elem:        &schema.Schema{Type: schema.TypeString},
+						},
+						"exclude": {
+							Type:        schema.TypeSet,
+							Optional:    true,
+							ForceNew:    true, // TODO: Remove this when updating is implemented
+							Description: "Array of ref names or patterns to exclude. The condition will not pass if any of these patterns match.",
+							Elem:        &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+			},
+			"rule_creation": {
```

Comment:
When I started implementing this, I played around with a nested `rules` field, which wrapped all the `rule_*` fields you can see below here. It proved to require a bunch of boilerplate code when implementing the actual Create/Read/Destroy logic, so I instead opted into moving them to the root, and instead using a `rule_*` naming convention for them. 

The `rule_*` addition might be overkill (and requires a little bit of parsing logic in the actual implementation), but since the schema is quite large, I think it's worth having it just to make it clear that these are the rules (a `creation` field doesn't really mean much, `rule_creation` is clearer IMO) . 

I'm open for feedback here though. Baking all of them into a nested structure is possible, just somewhat tedious.